File: csharp/AbstractFactory/ProductManager.cs
namespace AbstractFactory;

public class ProductManager
{
    private CrossCuttingConcernFactory _crossCuttingConcernFactory;

    private Logging _logging;
    private Caching _caching;

    public ProductManager(CrossCuttingConcernFactory crossCuttingConcernFactory)
    {
        _crossCuttingConcernFactory = crossCuttingConcernFactory;
        _logging = _crossCuttingConcernFactory.CreateLogger();
        _caching = _crossCuttingConcernFactory.CreateCaching();
    }

    public void GetAll()
    {
        _logging.Log("product log");
        _caching.Cache("product data");
        Console.WriteLine("Products listed.");
    }
}

File: java/abstract-factory/src/main/java/org/designpatterns/ProductManager.java
package org.designpatterns;

public class ProductManager {
    private final CrossCuttingConcernFactory crossCuttingConcernFactory;

    private final Logging logging;
    private final Caching caching;

    public ProductManager(CrossCuttingConcernFactory crossCuttingConcernFactory) {
        this.crossCuttingConcernFactory = crossCuttingConcernFactory;
        this.logging = crossCuttingConcernFactory.createLogger();
        this.caching = crossCuttingConcernFactory.createCaching();
    }

    public void getAll()
    {
        logging.Log("product log");
        caching.Cache("product data");
        System.out.println("Products listed");
    }
}